### PR TITLE
[MAILPOET-4236] Prevent activation of invalid automatic newsletters

### PR DESCRIPTION
**Pull request by johnolek**
This PR prevents a user from activating an automatic email (e.g. Welcome emails, post notification emails) from the email listing page if the newsletter is not valid for some reason. These reasons include:

1. There is no content
2. Emails are set to be sent through MSS and the newsletter does not include an unsubscribe link or URL
3. It is a re-engagement email and it doesn't include the `[link:subscription_re_engage_url]` shortcode
4. It is a re-engagement email and click tracking is disabled
5. It is a post notification email and it doesn't include an "Automated Latest Content" widget

Currently we are not checking for any of these things when toggling the status of a newsletter on a listing page like the following:

image

I decided to extract validation into its own service so that we can be more consistent wherever we need to validate if a newsletter is ready to be activated. We're not currently being consistent about what we're checking for where:

`\MailPoet\API\JSON\v1\Newsletters::setStatus` checks for reengagement requirements
`\MailPoet\API\JSON\v1\SendingQueue::validateNewsletter` checks for body content as well as unsubscribe links
https://github.com/mailpoet/mailpoet/blob/95857d74b97ff6443877c0811817d43920c6a430/mailpoet/assets/js/src/newsletter_editor/components/save.js#L316-L374 checks for empty content, re-engagement links, unsubscribe links, and ALC content.

## Ideas / Questions
1. Should we run this validation before creating a sending task for a newsletter? It's possible for a newsletter to go from a valid to invalid state after activation, for example if it's activated with a non-MSS sending method and then the sending method gets switched to MSS later.
2. I think it might make sense to replace some of the frontend validation code with something like an AJAX call to an endpoint that calls this new code. We currently have a lot of duplicated logic between the frontend and backend, with some validations only ever being handled in the frontend code. Keeping validation in both places will require more maintenance going forward and could lead to subtle issues where a newsletter could be considered valid in one context but not the other.

I definitely don't want to make the scope of this ticket too big, so if there are other things we would want to consider I think it would be best to create a followup ticket.

[MAILPOET-4236]

[MAILPOET-4236]: https://mailpoet.atlassian.net/browse/MAILPOET-4236?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by rodrigoprimo — CHANGES_REQUESTED** [review]
Good job, @johnolek! The code changes look code to me overall. The email validations that I tested are working, plus I cannot activate a welcome email anymore if the body is empty or missing an unsubscribe link (cc @veljkho I would suggest that when doing QA for this PR, you try to check as much as possible the places where we do email validation just to make sure everything is still working as expected).

Thanks for the detailed PR description!

I found one minor problem related to i18n that I believe we need to address before moving this forward. More details in an inline comment. I also left a few other inline comments sharing some thoughts and ideas that occurred to me while checking this PR.

One minor nitpick that I would like to mention as a suggestion for future PRs is that I would recommend avoiding commits containing multiple unrelated changes. For example, in my opinion, df4dfb5830df64cae36dc31f56c5021a2012391a could have been split into various smaller commits, and that would have made reviewing the changes a bit easier. 

> Should we run this validation before creating a sending task for a newsletter? It's possible for a newsletter to go from a valid to invalid state after activation, for example if it's activated with a non-MSS sending method and then the sending method gets switched to MSS later.

Good point. I think we should run this validation before a sending task, as you suggested. I would recommend checking with the rest of the team just to make sure we are not missing something, and then creating a new Jira ticket to address this.

> I think it might make sense to replace some of the frontend validation code with something like an AJAX call to an endpoint that calls this new code. 

+1 for the reasons that you mentioned

**Comment by rodrigoprimo** [comment]
I forgot to mention that there are some conflicts with master that needs to be solved before we can merge this PR.

**Comment by johnolek** [comment]
@rodrigoprimo Thank you for your review! I'll start working on some updates shortly.

I realized that I unfortunately forgot to submit a review of the code I started myself before I went on AFK. Those comments should now be published if you want to take a look at any of them.

> One minor nitpick that I would like to mention as a suggestion for future PRs is that I would recommend avoiding commits containing multiple unrelated changes. For example, in my opinion, https://github.com/mailpoet/mailpoet/commit/df4dfb5830df64cae36dc31f56c5021a2012391a could have been split into various smaller commits, and that would have made reviewing the changes a bit easier.

Thank you for pointing this out! I'm always happy to receive "nitpicky" feedback like this, and I don't even like calling it nitpicky because I think it diminishes the fact that smaller things can be very important.

Sometimes I find that I've moved forward with a few different changes that may not be all that related, but they might be hard to split into separate commits for one reason or another (like all relying on some piece of functionality that changed slightly). Do you have any recommendations for how to handle that kind of thing? 

The main thing I think I should focus on is simply committing more frequently and trying to maintain more focus on one particular thing so I don't end up with that kind of state in the first place. In the past I've worried about committing before I'm really certain that something is in the state I want it to be, but I recently started using fixup commits, which eliminates that issue because I know I have a relatively easy way of "editing" incomplete commits, but I've tended to avoid history-rewriting git operations in the past so any suggestions you might have, however basic, would be appreciated!

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (rodrigoprimo) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:67]: :+1:

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (rodrigoprimo) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:92]: I'm not familiar with this part of the code, but I found this file in case it helps:

https://github.com/mailpoet/mailpoet/blob/c50715698ce7137b72003fe742682ced355b5a16/mailpoet/assets/js/src/newsletter_editor/blocks/automatedLatestContent.js

I also noticed that we seem to have this type of block defined in several of our templates. Examples:

https://github.com/mailpoet/mailpoet/blob/9f790efbf0f032fc32e264dc33450286b362976e/mailpoet/lib/Config/PopulatorData/Templates/Avocado.php#L712
https://github.com/mailpoet/mailpoet/blob/9f790efbf0f032fc32e264dc33450286b362976e/mailpoet/lib/Config/PopulatorData/Templates/Charity.php#L824

That being said, I only skimmed through what I found, so I might be missing something.

File: mailpoet/tests/integration/API/JSON/v1/SendingQueueTest.php
Comment (rodrigoprimo) [reply mailpoet/tests/integration/API/JSON/v1/SendingQueueTest.php:161]: :+1:

File: mailpoet/tests/integration/_bootstrap.php
Comment (rodrigoprimo) [reply mailpoet/tests/integration/_bootstrap.php:170]: :+1:

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (rodrigoprimo) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:9]: I think its ok to use ValidationException here. I would have used one of the PHP built-in exceptions here since we don't need any of the methods that ValidationException provides, but I might be missing something regarding the downsides of using the built-in exceptions.

File: mailpoet/lib/API/JSON/v1/SendingQueue.php
Comment (rodrigoprimo) [inline mailpoet/lib/API/JSON/v1/SendingQueue.php:95]: Checking the code again after your comments, it occurred to me that there might be some redundancy here that I missed before. I'm saying this just but reading the code without testing, so I could be wrong. But now that we are using validate(), maybe we don't need to check for the subscriber limit anymore inside add() as validate already checks it (https://github.com/mailpoet/mailpoet/blob/3930d93c5a4a3ec5b27cfcc8d8d8b5fdf7bcd8f4/mailpoet/lib/API/JSON/v1/SendingQueue.php#L70)?

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (rodrigoprimo) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:32]: Yeah, I had a similar reaction about the name of this method as well, but couldn't think of anything better. I'm not sure about `readyToActivate`. This name makes sense to me when this method is used in Newsletters.php, but if I remember things correctly from my review last week, when this method is used in SendingQueue.php we are not activating an email after calling validate().

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (rodrigoprimo) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:34]: I don't know if we need to address this in this PR, but maybe a solution to the problem that you are raising, that would improve the user experience in my opinion, is to return all the errors that we find instead of just the first error.

**Comment by rodrigoprimo** [comment]
> Sometimes I find that I've moved forward with a few different changes that may not be all that related, but they might be hard to split into separate commits for one reason or another (like all relying on some piece of functionality that changed slightly). Do you have any recommendations for how to handle that kind of thing?
> The main thing I think I should focus on is simply committing more frequently and trying to maintain more focus on one particular thing so I don't end up with that kind of state in the first place.

The same thing definitely happens to me as well. Often times I find myself working on something and then realizing that I actually have multiple unrelated and uncommitted changes in my local branch. When that happens, I tend to pause and see if there are parts of the changes that could make for an individual commit. Sometimes it is easy to split those changes into multiple commits, sometimes it is not. I might use `git stash push` to stash just some specific files or even create local copies of a file if it contains changes that I believe should be split into multiple files. I create a copy, undo some of the changes in the original file, commit and then restore the other changes from the copy (there might be a better way to deal with changes in the same file that you want to add as separate commits).

> In the past I've worried about committing before I'm really certain that something is in the state I want it to be, but I recently started using fixup commits, which eliminates that issue because I know I have a relatively easy way of "editing" incomplete commits, but I've tended to avoid history-rewriting git operations in the past so any suggestions you might have, however basic, would be appreciated!

I use `git rebase --interactive` all the time to change commits when I'm working on a ticket and realize that something should be part of a commit that I already created or I should create a new commit before a commit that already exists. I don't have an issue with rewriting git history if I'm the only person working on a branch. I avoid rewriting git history if there are more people working on the same branch or if I already submitted the PR for a review. Let me know if there are issues with rewriting git history the way that I'm doing that I'm not aware of.

I hope the above helps. Happy to chat more about this.

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (johnolek) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:9]: I thought it might be nice to use a MailPoet-specific exception (i.e. one that extends `MailPoet/Exception`) in case we ever handle those separately. It probably doesn't matter too much in this case so I'll just leave it as is.

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (johnolek) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:34]: That occurred to me as well, though sometimes that approach can be extra noisy because one error might always imply another (e.g. if the email doesn't have any content, you might also see an error about it missing an unsubscribe link, which is kind of redundant). I'll leave this as-is for now because it doesn't seem like a major issue.

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (johnolek) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:92]: Thanks for tracking that down, not sure how I missed it myself!  I ended up trying to test this myself by creating a newsletter with an ALC block, and in that case what I see generated for the body in the DB looks like this:

```json
      {
        "type": "automatedLatestContentLayout",
        "withLayout": true,
        "amount": "5",
        "contentType": "post",
        "terms": [],
        "inclusionType": "include",
        "displayType": "excerpt",
        "titleFormat": "h1",
        "titleAlignment": "left",
        "titleIsLink": false,
        "imageFullWidth": false,
        "titlePosition": "abovePost",
        "featuredImagePosition": "alternate",
        "fullPostFeaturedImagePosition": "left",
        "showAuthor": "no",
...
```

I'm guessing perhaps there's a display option that changes the type from this to the version without `Layout`. In any case, there's no need I think to dig any further into this. Having the check for both definitely seems like what we want.

File: mailpoet/lib/API/JSON/v1/SendingQueue.php
Comment (johnolek) [reply mailpoet/lib/API/JSON/v1/SendingQueue.php:95]: I thought I made a separate comment about this but I don't think I actually ended up going into the detail I meant to, or I'm just not finding it. 

In any case, I'm finding myself wanting to go in the opposite direction and remove the subscriber check from the validator entirely. It feels wrong to have a `validateSubscriberLimit` method in the validator because it doesn't check anything specific to the newsletter itself. It's also nice to have a separate check and potential early return immediately inside `add` so we don't have to retrieve a newsletter first just to check if the subscriber limit has been hit. And if we did end up using this validator for the frontend side of things via an Ajax request, you would never be able to get meaningful validation errors anytime you would try to edit a newsletter when you're over the limit.

It really seems like an improper mixing of concerns and I think I'll remove it from the validator, though I'm certainly open to other ideas!

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (johnolek) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:32]: I'm going to leave this one as-is since it's clear enough and neither of us has a better idea.

**Comment by johnolek** [comment]
Thanks for your ideas about git workflows, @rodrigoprimo! I've been becoming more and more comfortable with interactive rebasing recently. People I've worked with before have had some pretty strong opinions about never (EVER) rewriting history, even for unpushed branches, and I'm finding I'm needing to unlearn some of that training.

I'm a big fan of PhpStorm's git integration myself, and I especially like how it gives you the option of excluding specific chunks of changes from a particular commit.

image
 
I also like its diff viewer in general.

Anyway, I definitely appreciate the ideas, and I'll continue to try to keep this conversation in mind any time I find myself writing another commit message that includes a list of unrelated changes.

File: mailpoet/lib/Newsletter/NewsletterValidator.php
Comment (rodrigoprimo) [reply mailpoet/lib/Newsletter/NewsletterValidator.php:34]: :+1:

File: mailpoet/lib/API/JSON/v1/SendingQueue.php
Comment (rodrigoprimo) [reply mailpoet/lib/API/JSON/v1/SendingQueue.php:95]: Sounds good to me. Fair point.

**Comment by rodrigoprimo** [comment]
> I'm a big fan of PhpStorm's git integration myself, and I especially like how it gives you the option of excluding specific chunks of changes from a particular commit.

@johnolek, I use Git in the command line, but I might try to use it inside PHPStorm next time I have to commit chunks of a file. Seems like a much easier way to do it than what I'm currently doing. Thanks for sharing it.

**Comment by johnolek** [comment]
@rodrigoprimo I believe this is ready for a final review. I think I addressed everything but of course please let me know if I missed something.

I did end up moving the subscriber limit check back out of the validator. I also refactored a few tests, and changed the way the `getServiceWithOverrides` helper works based on feedback I got from the P2 I posted. 

I did end up rebasing this branch, mostly to resolve conflicts with `master`, but I think that may have unlinked some of the comments you made. Sorry about that! So just one last `git` workflow question for now: for branches where a review has already started, is it okay to perform a merge commit with `master` if there are conflicts, or is it always best to rebase?

Thank you for your detailed review and for the good related discussions! And sorry if I ended up making this ticket more complicated than it needed to be!

**Review by rodrigoprimo — APPROVED** [review]
> I did end up moving the subscriber limit check back out of the validator. I also refactored a few tests, and changed the way the getServiceWithOverrides helper works based on feedback I got from the P2 I posted.

:+1: 

> I did end up rebasing this branch, mostly to resolve conflicts with master, but I think that may have unlinked some of the comments you made. Sorry about that! So just one last git workflow question for now: for branches where a review has already started, is it okay to perform a merge commit with master if there are conflicts, or is it always best to rebase?

In my opinion, in cases like this, it is always best to create merge commits instead of rebasing, but I'm not sure if everyone else on the team agrees with that. I prefer merging because it doesn't change the commit hashes and the dates of the commits. I believe rebasing makes the second review of a PR more complex as it is not immediately obvious which commits are new and which commits already existed, and there is also the chance that something inadvertently changed in the old commits, and there is no easy way to check that as far as I know.

> Thank you for your detailed review and for the good related discussions! And sorry if I ended up making this ticket more complicated than it needed to be!

There is nothing to be sorry for. We had some good discussions in this PR.

It looks good to me now, and I'm moving it to QA. Good job!

**Comment by johnolek** [comment]
@rodrigoprimo I definitely agree about preferring merges in these cases, and you make very good points in favor. 

I actually ran into a lot of annoying-to-resolve conflicts during my rebase that required resolution at multiple points instead of just as part of a single merge, and I ended up having to go back and fix a change I accidentally merged incorrectly. I ended up feeling a lot less confident about the changes myself and spent quite a bit of time reviewing the PR again myself to make sure there weren't any other issues.

This might be an interesting topic to bring up in a P2 as well, especially since I don't believe any of our guides or standards directly address the point of rebasing vs. merging.

Thank you again!

**Comment by rodrigoprimo** [comment]
> This might be an interesting topic to bring up in a P2 as well, especially since I don't believe any of our guides or standards directly address the point of rebasing vs. merging.

Yeah, I agree this is a good topic for a P2 post.

**Comment by rohankacharegs** [comment]
GlobalStep Verified

**Pull request merged**
